### ISSUE TITLE
Uniform predictions with DBN

### ISSUE THREAD

**Issue by hackingbear** [SkymindIO/deeplearning4j#105]
With the latest 0.0.3.3-SNAPSHOT, the problem with zeroed output layer weight matrix is gone. However the problem of uniform prediction still remained. I modified my previous dummy test to have two different inputs/outputs: vector of 100 mapped to 4th label set; vector of 200 mapped to 2nd label set. After the training, both input vector produced exactly the same (to all decimals) prediction, even though the values of the different labels for the same prediction look random.

The output
```
actual [0.0 ,0.0 ,0.0 ,1.0 ,0.0] vs predicted [0.044186138400415305 ,0.43372079239937705 ,0.044186138400415305 ,0.43372079239937705 ,0.044186138400415305]
actual [0.0 ,1.0 ,0.0 ,0.0 ,0.0] vs predicted [0.044186138400415305 ,0.43372079239937705 ,0.044186138400415305 ,0.43372079239937705 ,0.044186138400415305]
F1: 0.6
```

The test codes
```java
package org.deeplearning4j;

import org.apache.commons.math3.random.MersenneTwister;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.models.classifiers.dbn.DBN;
import org.deeplearning4j.models.featuredetectors.rbm.RBM;
import org.deeplearning4j.nn.WeightInit;
import org.deeplearning4j.nn.api.NeuralNetwork;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.nd4j.linalg.api.activation.Activations;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.lossfunctions.LossFunctions;

public class DBNDummyTest {
    public static void main(String... args) throws Exception {
        INDArray input = Nd4j.create(2, 614); // have to be at least two or else output layer gradient is a scalar and cause exception
        INDArray labels = Nd4j.create(2, 5);

        INDArray row0 = Nd4j.create(1, 614);
        row0.assign(100f);
        input.putRow(0, row0);
        labels.put(0, 3, 1); // set the 4th column

        INDArray row1 = Nd4j.create(1, 614);
        row1.assign(200f);
        input.putRow(1, row1);
        labels.put(1, 1, 1); // set the 2nd column

        DataSet trainingSet = new DataSet(input, labels);

        MersenneTwister gen = new MersenneTwister(123);

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
            .iterations(500)
            .weightInit(WeightInit.VI)
            .activationFunction(Activations.tanh())
            .visibleUnit(RBM.VisibleUnit.GAUSSIAN)
            .hiddenUnit(RBM.HiddenUnit.RECTIFIED)
            .constrainGradientToUnitNorm(false)
            .lossFunction(LossFunctions.LossFunction.XENT)
            .optimizationAlgo(NeuralNetwork.OptimizationAlgorithm.ITERATION_GRADIENT_DESCENT)
            .rng(gen)
            .learningRate(1e-2f)
            .nIn(trainingSet.numInputs()).nOut(trainingSet.numOutcomes()).list(2)
            .hiddenLayerSizes(new int[] {100})
            .override(new NeuralNetConfiguration.ConfOverride() {
                @Override
                public void override(int i, NeuralNetConfiguration.Builder builder) {
                    if (i == 1) {
                        builder.iterations(500);
                        builder.weightInit(WeightInit.ZERO);
                        builder.activationFunction(Activations.softMaxRows());
                        builder.lossFunction(LossFunctions.LossFunction.MCXENT);
                    }
                }
            }).build();

        DBN nn = new DBN.Builder().layerWiseConfiguration(conf).build();

        nn.fit(trainingSet);
        INDArray predict2 = nn.output(trainingSet.getFeatureMatrix());
        for (int i = 0; i < predict2.rows(); i++) {
            String actual = trainingSet.getLabels().getRow(i).toString().trim();
            String predicted = predict2.getRow(i).toString().trim();
            System.out.println("actual "+actual+" vs predicted " + predicted);
        }
//        for (int row = 0; row < nn.getOutputLayer().getW().rows(); row++) {
//            System.out.println(nn.getOutputLayer().getW().getRow(row).toString().trim());
//        }
        Evaluation eval = new Evaluation();
        eval.eval(trainingSet.getLabels(), predict2);
        System.out.println("F1: " + eval.f1());
    }
}
```

**Comment by hackingbear** [comment]
I'm able to update the test to the latest 0.0.3.3 with the new API and it produces exact predictions. One thing though, I have to set the iteration high. If I set it to something like 100 iterations, the results flicker among repeated runs (of the exact binary and data, with fixed RNG seed) because the label value differences are too small and easily impact by round off errors. This seems contradicting to advice in http://deeplearning4j.org/troubleshootingneuralnets.html  Is this normal for this very simple problem?

```
o.d.o.s.BaseOptimizer - Error at iteration 9999 was 0.10035656083660333
actual [0.0 ,0.0 ,0.0 ,1.0 ,0.0] vs predicted [0.022635427226149932 ,0.33985780191966286 ,0.022635427226149932 ,0.5922359164018873 ,0.022635427226149932]
actual [0.0 ,1.0 ,0.0 ,0.0 ,0.0] vs predicted [0.017826180473760635 ,0.6189595273749603 ,0.017826180473760635 ,0.3275619312037576 ,0.017826180473760635]
F1: 1.0
```

The updated test codes
```java
package org.deeplearning4j;

import org.apache.commons.math3.random.MersenneTwister;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.models.featuredetectors.rbm.RBM;
import org.deeplearning4j.nn.api.LayerFactory;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.layers.OutputLayer;
import org.deeplearning4j.nn.layers.factory.DefaultLayerFactory;
import org.deeplearning4j.nn.layers.factory.LayerFactories;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.nd4j.linalg.api.activation.Activations;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.lossfunctions.LossFunctions;

public class DBNDummyTest {
    public static void main(String... args) throws Exception {
        final int nSamples = 1000;
        int nFeatures = 10;
        INDArray input = Nd4j.create(nSamples, nFeatures); // have to be at least two or else output layer gradient is a scalar and cause exception
        INDArray labels = Nd4j.create(nSamples, 5);

        for (int i = 0; i < nSamples; i++) {
            INDArray row = Nd4j.create(1, nFeatures);
            if (i % 2 == 0) {
                row.assign(1f);
                labels.put(i, 3, 1); // set the 4th column
            } else {
                row.assign(2f);
                labels.put(i, 1, 1); // set the 2nd column
            }
            input.putRow(i, row);
        }

        DataSet trainingSet = new DataSet(input, labels);

        MersenneTwister gen = new MersenneTwister(123);
        final int iterations = 10000;
        LayerFactory layerFactory = LayerFactories.getFactory(RBM.class);
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
            .iterations(iterations)
            .weightInit(WeightInit.VI)
            .activationFunction(Activations.softMaxRows())
            .visibleUnit(RBM.VisibleUnit.BINARY)
            .hiddenUnit(RBM.HiddenUnit.SOFTMAX)
            .layerFactory(layerFactory)
            .constrainGradientToUnitNorm(true)
//            .lossFunction(LossFunctions.LossFunction.MCXENT)
            .optimizationAlgo(OptimizationAlgorithm.ITERATION_GRADIENT_DESCENT)
            .rng(gen)
            .learningRate(1e-3f)
            .nIn(trainingSet.numInputs()).nOut(trainingSet.numOutcomes()).list(2)
            .hiddenLayerSizes(new int[]{10})
            .override(new NeuralNetConfiguration.ConfOverride() {
                @Override
                public void override(int i, NeuralNetConfiguration.Builder builder) {
                    if (i == 1) {
                        builder.layerFactory(new DefaultLayerFactory(OutputLayer.class));
                        builder.iterations(iterations);
                        builder.weightInit(WeightInit.ZERO);
                        builder.activationFunction(Activations.softMaxRows());
                        builder.lossFunction(LossFunctions.LossFunction.MCXENT);
                    }
                }
            }).build();

        MultiLayerNetwork nn = new MultiLayerNetwork(conf);

        nn.fit(trainingSet);
        INDArray predict2 = nn.output(trainingSet.getFeatureMatrix());
        for (int i = 0; i < 2; i++) {
            String actual = trainingSet.getLabels().getRow(i).toString().trim();
            String predicted = predict2.getRow(i).toString().trim();
            System.out.println("actual "+actual+" vs predicted " + predicted);
        }
        Evaluation eval = new Evaluation();
        eval.eval(trainingSet.getLabels(), predict2);
        System.out.println("F1: " + eval.f1());
    }
}
```

**Comment by agibsonccc** [comment]
First normalize the data to zero mean and unit variance. Any neural net will produce weird output on data that isn't in between zero and 1. Especially with a DBN and binary data. If you still don't get signal changing the activation to tanh will help a lot as well.
For the hidden unit, I would recommend keeping it binary. 

Your configuration for the activation function is also off, softmax rows should be for a classifier only.

WRT your weight initialization, change that to be a smaller distribution. You can do this by changing the weight distribution to .DISTRIBUTION and specifying a distribution with .dist

**Comment by hackingbear** [comment]
Now with the recommended changes, I can get perfect prediction in 5 iterations! Though, why is this not a classifier? Only one of the output label is 1 and different label based on one of the two input feature vectors.

**Comment by agibsonccc** [comment]
Great! It's still a classifier. (Only the output layer though) Keep in mind the point of a DBN is to generate features using pretrained RBMs. The output layer (using multi class cross entropy as the objective) then has a softmax activation (which is essentially logistic/softmax regression if you use zero for the weights) Your goal then is to propagate a feature set through the RBMs such that the Logistic Regression classifier can learn features.

**Comment by agibsonccc** [comment]
I'm going to close this since it appears to be resolved. Please feel free to resume the conversation on the mailing list with tuning this particular example. I'd be more than glad to help you if you put this up with a github gist/repo. I'm aware the docs need to be improved and will continue working on this. I know it's a little daunting to tune neural nets and the suggestions I just gave you probably seem like they come from a black box or some sort of weird voodoo. There's definitely best practices with neural nets I hope to establish. I will continue working on examples and the like.
